pkgs/development/libraries/kerberos/heimdal.nix

```diff
@@ -134,6 +143,13 @@ stdenv.mkDerivation {
     export USER=nix-builder
   '';
 
+  postConfigure = ''
+    # Make build system include db_185.h where db.h v1 is expected
+    # Note: since ./configure doesn't allow to set the macro via a CLI flag, we
+    # have to edit config.h directly
+    echo "#define HAVE_DB_185_H 1" >> include/config.h
```

Comment:
this seems odd to say the least. linux and darwin are using the same db package and have the same db headers i believe.  linux doesn't define this, nor does it define `HAVE_DB_H` but somehow it builds? both linux and darwin are defining `HAVE_DBHEADER`.

i don't have really any insight into why linux is building and darwin failing but haven't looked too hard other than setting up some nix-shells and running
```
for x in unpackPhase patchPhase  $preConfigurePhases configurePhase; do runPhase $x; done
```
and looking around.

Reply:
On Linux, it looks like the configure check for HAVE_DB1 passes differently. Here is Linux:

```
checking /nix/store/bm8vh1ia3kydn42pkx9g428vhh8c6672-db-5.3.28-dev/include/db.h usability... yes
checking /nix/store/bm8vh1ia3kydn42pkx9g428vhh8c6672-db-5.3.28-dev/include/db.h presence... yes
checking for /nix/store/bm8vh1ia3kydn42pkx9g428vhh8c6672-db-5.3.28-dev/include/db.h... yes
checking for db_create... yes, in -ldb-5
checking for dbopen... no
checking for dbm_firstkey... no
```

Here's MacOS:

```
checking /nix/store/7dxg936k24n5g2iiiw00gnmxb35dvpyr-db-5.3.28-dev/include/db.h usability... yes
checking /nix/store/7dxg936k24n5g2iiiw00gnmxb35dvpyr-db-5.3.28-dev/include/db.h presence... yes
checking for /nix/store/7dxg936k24n5g2iiiw00gnmxb35dvpyr-db-5.3.28-dev/include/db.h... yes
checking for db_create... yes, in -ldb-5
checking for dbopen... yes
checking for dbm_firstkey... yes
checking lmdb.h usability... no
checking lmdb.h presence... no
checking for lmdb.h... no
```

The `dbopen` check sets `HAVE_DB1` which then makes an attempt to compile v1 implementation under `lib/hdb/db.c`.

I will try to understand why Linux is still not passing `dbopen` check and let you know.

Reply:
When `conftest` is linked, it fails to find the `dbopen` symbol on Linux:

```
| /* end confdefs.h.  */
| char dbopen (void);
| int
| main ()
| {
| dbopen()
|   ;
|   return 0;
| }
configure:17841: gcc  -o conftest -g -O2   -D_LARGE_FILES=  conftest.c  -ldb  -pthread -lpthread  >&5
conftest.c:46:9: warning: ISO C99 requires whitespace after the macro name
   46 | #define HAVE_LIBDB-5 1
      |         ^~~~~~~~~~
/nix/store/vcvhwiilizhijk7ywyn58p9l005n9sbn-binutils-2.43.1/bin/ld: /build/ccsUmbsT.o: in function `main':
/build/source/conftest.c:53:(.text.startup+0x5): undefined reference to `dbopen'
collect2: error: ld returned 1 exit status
configure:17841: $? = 1
```

I think it may be some difference in what is implicitly included in apple SDK vs the build environment for Linux that e.g. does not include v1 interface in `-ldb`? I am not sure, I wonder if someone with better knowledge about apple build environment could have an explanation.

Reply:
Here's what the conftest binary is linked to when built by nixpkgs@clang on darwin:

```
(ins)[nix-shell:~/src/nixpkgs]$ otool -L a.out
a.out:
	/nix/store/d0d4qn48b5wdb2368vgm9j93sq9ssblw-libresolv-64/lib/libresolv.9.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.60.1)
```

Reply:
Here is how `db_185.h` defines `dbopen`:

```
/* Re-define the user's dbopen calls. */
#define dbopen __db185_open
```

Indeed, when we inspect the db library with `ldd`, we don't see `dbopen` symbol exported; only `__db185_open`. Which results in linking failure when the compiled code refers to `dbopen`. If I replace `dbopen` references to `__db185_open`, then the linker succeeds.

On the other hand, libSystem in Darwin SDK seems to provide the symbol, so linking succeeds.

This looks like a bug in heimdal autoconf code, since they apparently fail to detect `dbopen` presence when it's just a macro.

Reply:
I've reported https://github.com/heimdal/heimdal/issues/1263 for upstream.